Initial QuestionAnsweringClient

heaths opened the pull request.

heaths committed:
Initial QuestionAnsweringClient

heaths committed:
Generate separate LLC for authoring client

heaths commented:
/azp run net - cognitivelanguage - ci

heaths committed:
Add cognitivelanugage to CODEOWNERS

heaths committed:
Define Query methods on QuestionAnsweringClient

heaths committed:
Remove QuestionAnsweringProjectsClient

heaths commented:
/azp run prepare-pipelines

heaths committed:
Add tests for QuestionAnsweringClient

heaths commented:
/azp run net - cognitivelanguage - tests

heaths committed:
Remove test resources from deployment template

We're using a set of static resources right now. Will add back once this service version is deployed generally.

heaths committed:
Remove master branch from ci.yml

heaths commented:
/azp run net - cognitivelanguage - tests

heaths committed:
Added README for Question Answering client library

pakrym commented on sdk/core/Azure.Core.TestFramework/src/RecordMatcher.cs:
```diff
@@ -242,6 +242,12 @@ private int CompareHeaderDictionaries(SortedDictionary<string, string[]> headers
                 var requestHeaderValues = header.Value;
                 var headerName = header.Key;
 
+                if (!_compareBodies && "Content-Length".Equals(headerName, StringComparison.OrdinalIgnoreCase))
```
You can achieve the same by adding `Content-Length` to ignored headers.

heaths replied:
But I believe this is a bug in the existing `RecordMatcher`. If we opted to not compare bodies, it shouldn't be comparing `Content-Length`.

heaths committed:
Resolve PR validation errors with URLs

heaths committed:
Resolve feedback and update recordings

The service deployed a fix for the confidenceScoreThreshold to truly be between 0 and 1.

heaths committed:
Add Azure client builder extensions

heaths committed:
Add more QuestionAnsweringClient samples

annatisch commented on sdk/cognitivelanguage/Azure.AI.Language.QuestionAnswering/README.md:
```diff
@@ -0,0 +1,213 @@
+
+### Install the package
+
+Install the Azure Key Vault secrets client library for .NET with [NuGet][nuget]:
```
Key Vault secrets? :)

annatisch commented on sdk/cognitivelanguage/Azure.AI.Language.QuestionAnswering/README.md:
````diff
@@ -0,0 +1,213 @@
+dotnet add package Azure.AI.Language.QuestionAnswering --prerelease
+```
+
+### Prerequisites
````
I would move the "Perquisites" section ahead of the "Install the package" section.

heaths replied:
This is our standard layout in the .NET repo. The prereqs is more about what you need for the subsequent examples, which follow.

maririos commented on sdk/cognitivelanguage/Azure.AI.Language.QuestionAnswering/src/QuestionAnsweringClient.cs:
```diff
@@ -0,0 +1,179 @@
+        /// Initializes a new instance of the <see cref="QuestionAnsweringClient"/> class.
+        /// </summary>
+        /// <param name="endpoint">The Question Answering endpoint on which to operate.</param>
+        /// <param name="credential">A <see cref="AzureKeyCredential"/> used to authenticate requests to the <paramref name="endpoint"/>.</param>
```
```suggestion
        /// <param name="credential">An <see cref="AzureKeyCredential"/> used to authenticate requests to the <paramref name="endpoint"/>.</param>
```

maririos commented on sdk/cognitivelanguage/Azure.AI.Language.QuestionAnswering/src/QuestionAnsweringClient.cs:
```diff
@@ -0,0 +1,179 @@
+        /// Initializes a new instance of the <see cref="QuestionAnsweringClient"/> class.
+        /// </summary>
+        /// <param name="endpoint">The Question Answering endpoint on which to operate.</param>
+        /// <param name="credential">A <see cref="AzureKeyCredential"/> used to authenticate requests to the <paramref name="endpoint"/>.</param>
```
```suggestion
        /// <param name="credential">An <see cref="AzureKeyCredential"/> used to authenticate requests to the <paramref name="endpoint"/>.</param>
```

maririos commented on sdk/cognitivelanguage/Azure.AI.Language.QuestionAnswering/src/QuestionAnsweringClient.cs:
```diff
@@ -0,0 +1,179 @@
+        /// <summary>
+        /// Get the service endpoint for this client.
+        /// </summary>
+        public virtual Uri Endpoint { get; }
```
ohh interesting. I didn't know we were exposing the `endpoint` in the client. is this a new guideline?

heaths replied:
We did for Key Vault and Search. The idea was that it's easy to tell from a client with which endpoint you're communicating. It's not a guideline as much as a convenience. Tagging @KrzysztofCwalina and @tg-msft in case they want me to remove this, but it's been through approval in those aforementioned libraries. In fact, @tg-msft was the one who added it to Search first.

maririos replied:
ohh cool, thanks!

maririos commented on sdk/cognitivelanguage/Azure.AI.Language.QuestionAnswering/src/QuestionAnsweringClient.cs:
```diff
@@ -0,0 +1,179 @@
+
+        /// <summary>Answers the specified question using your knowledgebase.</summary>
+        /// <param name="projectName">The name of the project to use.</param>
+        /// <param name="options">The question to answer.</param>
```
my understanding is that there are other properties that can be set in here right? Consider saying something along those lines too

maririos commented on sdk/cognitivelanguage/Azure.AI.Language.QuestionAnswering/src/QuestionAnsweringClient.cs:
```diff
@@ -0,0 +1,179 @@
+        /// <summary>Answers the specified question using your knowledgebase.</summary>
+        /// <param name="projectName">The name of the project to use.</param>
+        /// <param name="options">The question to answer.</param>
+        /// <param name="deploymentName">The optional deployment name of the project to use. If not specified, the production knowledgebase will be queried.</param>
```
what does the production knowledge base is?

heaths replied:
I'll reword. The deployments are like slots in other services, with defaults "test" and "prod" slots available.

maririos commented on sdk/cognitivelanguage/Azure.AI.Language.QuestionAnswering/src/QuestionAnsweringClientOptions.cs:
```diff
@@ -0,0 +1,37 @@
+
+namespace Azure.AI.Language.QuestionAnswering
+{
+    /// <summary> Client options for QuestionAnsweringClient. </summary>
```
```suggestion
    /// <summary> Client options for the <see cref="QuestionAnsweringClient"/>. </summary>
```

heaths replied:
I may not actually be able to change this. This was copied from generated code and I forget why off hand, but if/when it's generated again I won't be able to change it. I can file a bug on the generator, though.

heaths replied:
I changed it for now, at least. IIRC, once I add authoring support back in I may lose that ability, but I'll address is then if I can.

heaths commented on sdk/cognitivelanguage/Azure.AI.Language.QuestionAnswering/README.md:
```diff
@@ -0,0 +1,213 @@
+* An [Azure subscription][azure_subscription]
+* An existing Question Answering resource
+
+> Note: the new unified Cognitive Language Services are not currently available for deployment.
```
No one outside the service team can deploy a QA service that supports this api-version yet.

maririos replied:
I see. but the user can still create just a single Q&A resource, which is what you are showing, right?

heaths replied:
They can't even use this SDK right now. This is more of a placeholder until the service supports a public preview.

heaths commented on sdk/cognitivelanguage/Azure.AI.Language.QuestionAnswering/README.md:
```diff
@@ -0,0 +1,213 @@
+
+> Note: the new unified Cognitive Language Services are not currently available for deployment.
+
+### Authenticate the client
```
Correct

heaths commented on sdk/cognitivelanguage/Azure.AI.Language.QuestionAnswering/README.md:
````diff
@@ -0,0 +1,213 @@
+
+The only input required to a ask a question using a knowledgebase is just the question itself:
+
+```C# Snippet:QuestionAnsweringClient_QueryKnowledgebase
````
Is that true @KrzysztofCwalina? I hadn't heard that but more than happy to use async samples instead.

heaths replied:
@tg-msft isn't aware of such a change, but we can always change them (probably change all the READMEs) if that becomes the case. I think it shows best practices.

maririos replied:
I guess I am the off one :P 
Thanks for checking

heaths commented on sdk/cognitivelanguage/Azure.AI.Language.QuestionAnswering/README.md:
````diff
@@ -0,0 +1,213 @@
+```C# Snippet:QuestionAnsweringClient_Chat
+KnowledgebaseQueryOptions options = new KnowledgebaseQueryOptions("How long should charging take?")
+{
+    Context = new KnowledgebaseAnswerRequestContext(previousAnswer.Id.Value)
````
I had named it in hopes that it was an obvious reference to the previous answer, but since that wasn't clear I will change it somehow to clarify. It was supposed to be like a follow-up from the previous sample. Perhaps I can simulate something of that to feed into this question.

heaths commented on sdk/cognitivelanguage/Azure.AI.Language.QuestionAnswering/README.md:
```diff
@@ -0,0 +1,213 @@
+{
+    "error": {
+    "code": "BadArgument",
+    "message": "Please verify azure search service is up, restart the WebApp and try again"
```
@rokulka this message would be good to improve. In this case, the knowledgebase was non-existent so a clearer message would be great.

rokulka replied:
@heaths  - Yes, I will be going through and fixing all the error messages as my next task item. We will change this appropriately.

heaths commented on sdk/cognitivelanguage/Azure.AI.Language.QuestionAnswering/README.md:
```diff
@@ -0,0 +1,213 @@
+
+### Ask a question
+
+The only input required to a ask a question using a knowledgebase is just the question itself:
```
I agree it probably should be "knowledge base" in writing everywhere, but that also conflicts with "knowledgebase" (as opposed to, say, "knowledgeBase") throughout the swagger. @annatisch @rokulka do you see any problems with us in written language to use "knowledge base" everywhere consistently unless, of course, referring to a parameter or property name?

heaths replied:
I opened https://github.com/Azure/azure-sdk-for-net/issues/22510 to track.

/cc @annatisch

heaths commented on sdk/cognitivelanguage/Azure.AI.Language.QuestionAnswering/src/QuestionAnsweringClient.cs:
```diff
@@ -0,0 +1,179 @@
+        /// <param name="cancellationToken">An optional <see cref="CancellationToken"/> to cancel the request.</param>
+        /// <exception cref="ArgumentNullException"><paramref name="projectName"/> or <paramref name="options"/> is null.</exception>
+        /// <exception cref="RequestFailedException">The service returned an error. The exception contains details of the service error.</exception>
+        public virtual async Task<Response<KnowledgebaseAnswers>> QueryKnowledgebaseAsync(string projectName, KnowledgebaseQueryOptions options, string deploymentName = null, CancellationToken cancellationToken = default)
```
@KrzysztofCwalina @tg-msft @schaabs it came up in review that we should move required parameters out of `KnowledgebaseQueryOptions` and put them as required parameters, which is either `string question` or `int qnaId`. Agreed? How does that differ from `EncryptParameters` for Key Vault, though?

Also, since not all languages support overloads, I propose adding just a `GetQuestionAnswer(string projectName, int qnaId, string deploymentName = null)`. when a `qnaId` is passed, it's the only option that matters.

/cc @annatisch @rokulka

heaths replied:
Also adding @johanste since this would affect Python as well, which doesn't support overloads.

heaths replied:
Tracking in https://github.com/Azure/azure-sdk-for-net/issues/22511

heaths committed:
Resolve PR feedback

heaths committed:
Rebase on main for codegen changes
